src/Orleans.Core/Messaging/GatewayConnection.cs

```diff
@@ -250,28 +250,48 @@ protected override void OnGetSendingSocketFailure(Message msg, string error)
         protected override void OnMessageSerializationFailure(Message msg, Exception exc)
         {
-            // we only get here if we failed to serialise the msg (or any other catastrophic failure).
-            // Request msg fails to serialise on the sending silo, so we just enqueue a rejection msg.
-            Log.Warn(ErrorCode.ProxyClient_SerializationError, $"Unexpected error serializing message to gateway {Address}.", exc);
-            FailMessage(msg, $"Unexpected error serializing message to gateway {Address}. {exc}");
-            if (msg.Direction == Message.Directions.Request || msg.Direction == Message.Directions.OneWay)
+            // we only get here if we failed to serialize the msg (or any other catastrophic failure).
+            // Request msg fails to serialize on the sender, so we just enqueue a rejection msg.
+            // Response msg fails to serialize on the responding silo, so we try to send an error response back.
+            if (this.Log.IsEnabled(LogLevel.Warning))
             {
-                return;
+                this.Log.LogWarning(
+                    (int)ErrorCode.ProxyClient_SerializationError,
+                    "Unexpected error serializing message {Message}: {Exception}",
+                    msg,
+                    exc);
             }
 
-            // Response msg fails to serialize on the responding silo, so we try to send an error response back.
-            // if we failed sending an original response, turn the response body into an error and reply with it.
-            msg.Result = Message.ResponseTypes.Error;
-            msg.BodyObject = Response.ExceptionResponse(exc);
-            try
+            msg.ReleaseBodyAndHeaderBuffers();
+            MessagingStatisticsGroup.OnFailedSentMessage(msg);
+
+            var retryCount = msg.RetryCount ?? 0;
+
+            if (msg.Direction == Message.Directions.Request)
             {
-                MsgCenter.SendMessage(msg);
+                this.MsgCenter.RejectMessage(msg, $"Unable to serialize message. Encountered exception: {exc?.GetType()}: {exc?.Message}", exc);
             }
-            catch (Exception ex)
+            else if (msg.Direction == Message.Directions.Response && retryCount < 1)
             {
-                // If we still can't serialize, drop the message on the floor
-                Log.Warn(ErrorCode.ProxyClient_DroppingMsg, "Unable to serialize message - DROPPING " + msg, ex);
-                msg.ReleaseBodyAndHeaderBuffers();
+                // if we failed sending an original response, turn the response body into an error and reply with it.
+                // unless we have already tried sending the response multiple times.
+                msg.Result = Message.ResponseTypes.Error;
+                msg.BodyObject = Response.ExceptionResponse(exc);
+                msg.RetryCount = retryCount + 1;
+                this.MsgCenter.SendMessage(msg);
```

Comment:
what if serialization on this msg failed as well? do we want to add a try catch here, if this exception failed to serialize, we flatten it to a string, or a minimal Exception to return to the caller, stating serialization issue happened?

Reply:
SendMessage just enqueues the message, so no serialization happens at that stage. If this also fails to be serialized then something is very wrong.
In the case that we're failing while sending a request message, the rejection will be handled locally and therefore won't be serialized. For response messages, the entire message body is being replaced by a different exception, so if that doesn't work then we have big issues (catastrophe).

If the message still fails then it could be a RequestContext issue, but that means that we won't even be able to deserialize message headers (so we have no information to go by)

Longer term, I'd like to move RequestContext out of the headers for this very reason.